In [325]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import statistics  as sts
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import cross_val_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report

### Carregamento da base dados pelo Drive

In [326]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [327]:
path = '/content/drive/MyDrive/Inteligência Artificial /Machine Learning/'
arquivo = "Hotel Reservations.csv";
df=pd.read_csv(path+arquivo)
df

Booking_ID  no_of_adults  no_of_children  no_of_weekend_nights  \
0       INN00001             2               0                     1   
1       INN00002             2               0                     2   
2       INN00003             1               0                     2   
3       INN00004             2               0                     0   
4       INN00005             2               0                     1   
...          ...           ...             ...                   ...   
36270   INN36271             3               0                     2   
36271   INN36272             2               0                     1   
36272   INN36273             2               0                     2   
36273   INN36274             2               0                     0   
36274   INN36275             2               0                     1   

       no_of_week_nights type_of_meal_plan  required_car_parking_space  \
0                      2       Meal Plan 1                           0   
1                      3      Not Selected                           0   
2                      1       Meal Plan 1                           0   
3                      2       Meal Plan 1                           0   
4                      1      Not Selected                           0   
...                  ...               ...                         ...   
36270                  6       Meal Plan 1                           0   
36271                  3       Meal Plan 1                           0   
36272                  6       Meal Plan 1                           0   
36273                  3      Not Selected                           0   
36274                  2       Meal Plan 1                           0   

      room_type_reserved  lead_time  arrival_year  arrival_month  \
0            Room_Type 1        224          2017             10   
1            Room_Type 1          5          2018             11   
2            Room_Type 1          1          2018              2   
3            Room_Type 1        211          2018              5   
4            Room_Type 1         48          2018              4   
...                  ...        ...           ...            ...   
36270        Room_Type 4         85          2018              8   
36271        Room_Type 1        228          2018             10   
36272        Room_Type 1        148          2018              7   
36273        Room_Type 1         63          2018              4   
36274        Room_Type 1        207          2018             12   

       arrival_date market_segment_type  repeated_guest  \
0                 2             Offline               0   
1                 6              Online               0   
2                28              Online               0   
3                20              Online               0   
4                11              Online               0   
...             ...                 ...             ...   
36270             3              Online               0   
36271            17              Online               0   
36272             1              Online               0   
36273            21              Online               0   
36274            30             Offline               0   

       no_of_previous_cancellations  no_of_previous_bookings_not_canceled  \
0                                 0                                     0   
1                                 0                                     0   
2                                 0                                     0   
3                                 0                                     0   
4                                 0                                     0   
...                             ...                                   ...   
36270                             0                                     0   
36271                             0                                     0   
36272                            

In [328]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36275 entries, 0 to 36274
Data columns (total 19 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   Booking_ID                            36275 non-null  object 
 1   no_of_adults                          36275 non-null  int64  
 2   no_of_children                        36275 non-null  int64  
 3   no_of_weekend_nights                  36275 non-null  int64  
 4   no_of_week_nights                     36275 non-null  int64  
 5   type_of_meal_plan                     36275 non-null  object 
 6   required_car_parking_space            36275 non-null  int64  
 7   room_type_reserved                    36275 non-null  object 
 8   lead_time                             36275 non-null  int64  
 9   arrival_year                          36275 non-null  int64  
 10  arrival_month                         36275 non-null  int64  
 11  arrival_date   

In [329]:
df.describe()

no_of_adults  no_of_children  no_of_weekend_nights  no_of_week_nights  \
count  36275.000000    36275.000000          36275.000000       36275.000000   
mean       1.844962        0.105279              0.810724           2.204300   
std        0.518715        0.402648              0.870644           1.410905   
min        0.000000        0.000000              0.000000           0.000000   
25%        2.000000        0.000000              0.000000           1.000000   
50%        2.000000        0.000000              1.000000           2.000000   
75%        2.000000        0.000000              2.000000           3.000000   
max        4.000000       10.000000              7.000000          17.000000   

       required_car_parking_space     lead_time  arrival_year  arrival_month  \
count                36275.000000  36275.000000  36275.000000   36275.000000   
mean                     0.030986     85.232557   2017.820427       7.423653   
std                      0.173281     85.930817      0.383836       3.069894   
min                      0.000000      0.000000   2017.000000       1.000000   
25%                      0.000000     17.000000   2018.000000       5.000000   
50%                      0.000000     57.000000   2018.000000       8.000000   
75%                      0.000000    126.000000   2018.000000      10.000000   
max                      1.000000    443.000000   2018.000000      12.000000   

       arrival_date  repeated_guest  no_of_previous_cancellations  \
count  36275.000000    36275.000000                  36275.000000   
mean      15.596995        0.025637                      0.023349   
std        8.740447        0.158053                      0.368331   
min        1.000000        0.000000                      0.000000   
25%        8.000000        0.000000                      0.000000   
50%       16.000000        0.000000                      0.000000   
75%       23.000000        0.000000                      0.000000   
max       31.000000        1.000000                     13.000000   

       no_of_previous_bookings_not_canceled  avg_price_per_room  \
count                          36275.000000        36275.000000   
mean                               0.153411          103.423539   
std                                1.754171           35.089424   
min                                0.000000            0.000000   
25%                                0.000000           80.300000   
50%                                0.000000           99.450000   
75%                                0.000000          120.000000   
max                               58.000000          540.000000   

       no_of_special_requests  
count            36275.000000  
mean                 0.619655  
std                  0.786236  
min                  0.000000  
25%                  0.000000  
50%                  0.000000  
75%                  1.000000  
max                  5.000000

In [330]:
df.shape

(36275, 19)

In [331]:
df.isnull().sum()

Booking_ID                              0
no_of_adults                            0
no_of_children                          0
no_of_weekend_nights                    0
no_of_week_nights                       0
type_of_meal_plan                       0
required_car_parking_space              0
room_type_reserved                      0
lead_time                               0
arrival_year                            0
arrival_month                           0
arrival_date                            0
market_segment_type                     0
repeated_guest                          0
no_of_previous_cancellations            0
no_of_previous_bookings_not_canceled    0
avg_price_per_room                      0
no_of_special_requests                  0
booking_status                          0
dtype: int64

## 2) Pré-processamento

### Remoção ID feature e verificação de dados faltantes

In [332]:
df.drop(['Booking_ID'],axis=1, inplace=True)


### Encoding Categorical Features
* https://www.kaggle.com/code/alexisbcook/categorical-variables/tutorial
* https://towardsdatascience.com/categorical-encoding-using-label-encoding-and-one-hot-encoder-911ef77fb5bd

* Atributos com apenas dois valores
* Classe LabelEncoder

In [333]:
from sklearn.preprocessing import LabelEncoder

Verifcando a quantidade de possíveis valores que as colunas categóricas podem ter

In [334]:
df['booking_status'].unique()

array(['Not_Canceled', 'Canceled'], dtype=object)

In [335]:
df['arrival_year'].unique()

array([2017, 2018])

In [336]:
labelencoder = LabelEncoder()
df['booking_status'] = labelencoder.fit_transform(df["booking_status"])
df['arrival_year'] = labelencoder.fit_transform(df['arrival_year'])

* Atributos com mais de dois valores
* Função get_dummies()

In [337]:
df['type_of_meal_plan'].unique()

array(['Meal Plan 1', 'Not Selected', 'Meal Plan 2', 'Meal Plan 3'],
      dtype=object)

In [338]:
df['room_type_reserved'].unique()

array(['Room_Type 1', 'Room_Type 4', 'Room_Type 2', 'Room_Type 6',
       'Room_Type 5', 'Room_Type 7', 'Room_Type 3'], dtype=object)

In [339]:
df['market_segment_type'].unique()

array(['Offline', 'Online', 'Corporate', 'Aviation', 'Complementary'],
      dtype=object)

In [340]:
df = pd.get_dummies(df, columns=['type_of_meal_plan','room_type_reserved','market_segment_type','arrival_month','arrival_date'])


### Normalização dos dados

In [341]:
std=StandardScaler()
columns = ['no_of_adults','no_of_children','no_of_weekend_nights','no_of_week_nights','lead_time','no_of_previous_cancellations','no_of_previous_bookings_not_canceled',
'avg_price_per_room','no_of_special_requests']
df[['no_of_adults','no_of_children','no_of_weekend_nights','no_of_week_nights','lead_time','no_of_previous_cancellations','no_of_previous_bookings_not_canceled',
'avg_price_per_room','no_of_special_requests']] =\
std.fit_transform(df[['no_of_adults','no_of_children','no_of_weekend_nights','no_of_week_nights','lead_time','no_of_previous_cancellations','no_of_previous_bookings_not_canceled',
'avg_price_per_room','no_of_special_requests']])

In [342]:
df

no_of_adults  no_of_children  no_of_weekend_nights  no_of_week_nights  \
0          0.298893        -0.26147              0.217401          -0.144803   
1          0.298893        -0.26147              1.365993           0.563972   
2         -1.628975        -0.26147              1.365993          -0.853578   
3          0.298893        -0.26147             -0.931190          -0.144803   
4          0.298893        -0.26147              0.217401          -0.853578   
...             ...             ...                   ...                ...   
36270      2.226761        -0.26147              1.365993           2.690296   
36271      0.298893        -0.26147              0.217401           0.563972   
36272      0.298893        -0.26147              1.365993           2.690296   
36273      0.298893        -0.26147             -0.931190           0.563972   
36274      0.298893        -0.26147              0.217401          -0.144803   

       required_car_parking_space  lead_time  arrival_year  repeated_guest  \
0                               0   1.614896             0               0   
1                               0  -0.933701             1               0   
2                               0  -0.980250             1               0   
3                               0   1.463610             1               0   
4                               0  -0.433291             1               0   
...                           ...        ...           ...             ...   
36270                           0  -0.002706             1               0   
36271                           0   1.661446             1               0   
36272                           0   0.730452             1               0   
36273                           0  -0.258730             1               0   
36274                           0   1.417060             1               0   

       no_of_previous_cancellations  no_of_previous_bookings_not_canceled  \
0                         -0.063393                             -0.087456   
1                         -0.063393                             -0.087456   
2                         -0.063393                             -0.087456   
3                         -0.063393                             -0.087456   
4                         -0.063393                             -0.087456   
...                             ...                                   ...   
36270                     -0.063393                             -0.087456   
36271                     -0.063393                             -0.087456   
36272                     -0.063393                             -0.087456   
36273                     -0.063393                             -0.087456   
36274                     -0.063393                             -0.087456   

       ...  arrival_date_22  arrival_date_23  arrival_date_24  \
0      ...                0                0                0   
1      ...                0                0                0   
2      ...                0                0                0   
3      ...                0                0                0   
4      ...                0                0                0   
...    ...              ...              ...              ...   
36270  ...                0                0                0   
36271  ...                0                0                0   
36272  ...                0                0                0   
36273  ...                0                0                0   
36274  ...                0                0                0   

       arrival_date_25  arrival_date_26  arrival_date_27  arrival_date_28  \
0                    0                0                0                0   
1                    0                0                0                0   
2                    0                0                0                1   
3                    0                0                0                0   
4                 

## 3) Algoritmos de Classificação
* Árvore de Decisão
* KNN

### Separação entre treino e teste (70% e 30%)

In [343]:
X = df.drop(['booking_status'], axis=1).values
y = df['booking_status'].values

In [344]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

### Decision Tree

In [345]:

DT = DecisionTreeClassifier()
DT.fit(X_train, y_train)
y_pred = DT.predict(X_test)

DT_score = DT.score(X_train, y_train)
DT_test = DT.score(X_test, y_test)

cm = confusion_matrix(y_test,y_pred)
print('Training Score',DT_score)
print('Testing Score \n',DT_test)
print(cm)

Training Score 0.994328922495274
Testing Score 
 0.8667646788569329
[[2811  728]
 [ 722 6622]]


### KNN Classifier

In [346]:
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train, y_train)
y_pred = knn.predict(X_test)

train_score = knn.score(X_train,y_train)
test_score = knn.score(X_test, y_test)

cm = confusion_matrix(y_test,y_pred)
print('Training Score',train_score)
print('Testing Score \n',test_score)
print(cm)

Training Score 0.8933522369250158
Testing Score 
 0.848387393182027
[[2613  926]
 [ 724 6620]]


## 4) validação cruzada

- https://drigols.medium.com/introdu%C3%A7%C3%A3o-a-valida%C3%A7%C3%A3o-cruzada-k-fold-2a6bced32a90
- https://dataml.com.br/validacao-cruzada-aninhada-com-scikit-learn/

In [347]:
from sklearn.model_selection import StratifiedKFold

#### Decision Tree - Cross-Validation

In [348]:
# Create StratifiedKFold object.
skf = StratifiedKFold(n_splits=10, shuffle=True, random_state=1)
lst_accu_stratified = []

for train_index, test_index in skf.split(X, y):
    x_train_fold, x_test_fold = X[train_index], X[test_index]
    y_train_fold, y_test_fold = y[train_index], y[test_index]
    DT.fit(x_train_fold, y_train_fold)
    lst_accu_stratified.append(DT.score(x_test_fold, y_test_fold))

# Print the output.
print('Lista de ACC:', lst_accu_stratified)
print('\nMaior ACC:',
      max(lst_accu_stratified)*100, '%')
print('\nMenor ACC:',
      min(lst_accu_stratified)*100, '%')
print('\nMédia ACC:',
      sts.mean(lst_accu_stratified)*100, '%')
print('\nDesvio Padrão:', sts.stdev(lst_accu_stratified))

Lista de ACC: [0.8806504961411246, 0.869625137816979, 0.8770672546857773, 0.8861631753031973, 0.8756890848952591, 0.8682106424041908, 0.873724841466777, 0.8709677419354839, 0.87648194099807, 0.8753791011855528]

Maior ACC: 88.61631753031973 %

Menor ACC: 86.82106424041908 %

Média ACC: 87.53959416832411 %

Desvio Padrão: 0.005322225855565388


#### KNN - Cross-Validation

In [349]:
# Create StratifiedKFold object.
skf = StratifiedKFold(n_splits=10, shuffle=True, random_state=1)
lst_accu_stratified = []

for train_index, test_index in skf.split(X, y):
    x_train_fold, x_test_fold = X[train_index], X[test_index]
    y_train_fold, y_test_fold = y[train_index], y[test_index]
    knn.fit(x_train_fold, y_train_fold)
    lst_accu_stratified.append(knn.score(x_test_fold, y_test_fold))

# Print the output.
print('Lista de ACC:', lst_accu_stratified)
print('\nMaior ACC:',
      max(lst_accu_stratified)*100, '%')
print('\nMenor ACC:',
      min(lst_accu_stratified)*100, '%')
print('\nMédia ACC:',
      sts.mean(lst_accu_stratified)*100, '%')
print('\nDesvio Padrão:', sts.stdev(lst_accu_stratified))

Lista de ACC: [0.8472987872105844, 0.847574421168688, 0.8453693495038589, 0.8522601984564498, 0.8539140022050716, 0.8414667769506479, 0.8409153570443894, 0.8505652054039151, 0.8486352357320099, 0.8497380755445272]

Maior ACC: 85.39140022050717 %

Menor ACC: 84.09153570443894 %

Média ACC: 84.77737409220141 %

Desvio Padrão: 0.004256835014105341


## 5) Balanceamento das classes
- https://medium.com/analytics-vidhya/undersampling-and-oversampling-an-old-and-a-new-approach-4f984a0e8392
* Abordagem SMOTE

In [350]:
df['booking_status'].value_counts()

1    24390
0    11885
Name: booking_status, dtype: int64

In [351]:
from imblearn.over_sampling import SMOTE

#### Decision Tree - Balanceamento e Validação Cruzada

In [352]:
# Create StratifiedKFold object.
skf = StratifiedKFold(n_splits=10, shuffle=True, random_state=1)
lst_accu_stratified = []

for train_index, test_index in skf.split(X, y):
    x_train_fold, x_test_fold = X[train_index], X[test_index]
    y_train_fold, y_test_fold = y[train_index], y[test_index]
    sm = SMOTE()
    x_train_oversampled, y_train_oversampled = sm.fit_resample(x_train_fold, y_train_fold)
    DT.fit(x_train_oversampled, y_train_oversampled)
    lst_accu_stratified.append(DT.score(x_test_fold, y_test_fold))

# Print the output.
print('Lista de ACC:', lst_accu_stratified)
print('\nMaior ACC:',
      max(lst_accu_stratified)*100, '%')
print('\nMenor ACC:',
      min(lst_accu_stratified)*100, '%')
print('\nMédia ACC:',
      sts.mean(lst_accu_stratified)*100, '%')
print('\nDesvio Padrão:', sts.stdev(lst_accu_stratified))

Lista de ACC: [0.8781697905181919, 0.8690738699007717, 0.8715545755237045, 0.88257993384785, 0.8798235942668137, 0.8577336641852771, 0.8789633305762338, 0.8833746898263027, 0.869313482216708, 0.872897711607389]

Maior ACC: 88.33746898263027 %

Menor ACC: 85.7733664185277 %

Média ACC: 87.43484642469242 %

Desvio Padrão: 0.007847440364806784


#### KNN - Balanceamento e Validação Cruzada

In [353]:
# Create StratifiedKFold object.
skf = StratifiedKFold(n_splits=10, shuffle=True, random_state=1)
lst_accu_stratified = []

for train_index, test_index in skf.split(X, y):
    x_train_fold, x_test_fold = X[train_index], X[test_index]
    y_train_fold, y_test_fold = y[train_index], y[test_index]
    sm = SMOTE()
    x_train_oversampled, y_train_oversampled = sm.fit_resample(x_train_fold, y_train_fold)
    knn.fit(x_train_oversampled, y_train_oversampled)
    lst_accu_stratified.append(knn.score(x_test_fold, y_test_fold))

# Print the output.
print('Lista de ACC:', lst_accu_stratified)
print('\nMaior ACC:',
      max(lst_accu_stratified)*100, '%')
print('\nMenor ACC:',
      min(lst_accu_stratified)*100, '%')
print('\nMédia ACC:',
      sts.mean(lst_accu_stratified)*100, '%')
print('\nDesvio Padrão:', sts.stdev(lst_accu_stratified))

Lista de ACC: [0.8189084895259096, 0.8180815876515987, 0.8125689084895259, 0.8211135611907387, 0.820837927232635, 0.8158257513096223, 0.8025916735594155, 0.8169285911221396, 0.8238213399503722, 0.8205128205128205]

Maior ACC: 82.38213399503722 %

Menor ACC: 80.25916735594156 %

Média ACC: 81.71190650544779 %

Desvio Padrão: 0.005998244734371526


#Conclusão

O maior problema encontrado na base de dados foi a forma de utilização dos atributos arrival_date, arrival_month, e arrival_year. Os algoritmos KNN e Árvore de Decisão foram aplicados separadamente para cada um dos cenários abaixo:

1º Retirada (drop) desses atributos, assim como feito com ID
    
    Aqui a acurácia máxima (pós balanceamento validação cruzada) e o score de teste se mantiveram em 85.60% para AD e 85.24% para KNN.
    AD - Score
    Treinamento: 0.9922416509136736
    Teste:       0.8495819167508959
    KNN - Score
    Treinamento: 0.8946518588531821
    Teste:       0.8457226867591656


    Aqui houve uma diminuição do score de predição em ambos os algoritmos.

2º Definindo arrival_year como uma categorical feature de dois valores (2017 e
   2018), e arrival_date e arrival_month como uma categorical feature com mais de 2 valores.

    Aqui a acurácia máxima (pós balanceamento validação cruzada) e o score de teste se mantiveram em 88.33% para AD e 82.38% para KNN.

    AD - Score
    Treinamento: 0.9946439823566477
    Teste: 0.8669484517136818
    KNN - Score
    Treinamento: 0.8944549464398236
    Teste: 0.8459983460442893



3º Definindo arrival_year, arrival_month e arrival_date como atributos normalizados:

    Aqui a acurácia máxima (pós balanceamento validação cruzada) e o score de teste se mantiveram em 87.87% para AD e 84.20% para KNN.

    AD - Score
    Treinamento: 0.9945258349086327
    Teste 0.8627216760084535
    KNN - Score
    Treinamento: 0.8944943289224953
    Teste: 0.8406689331985666

  
Possuindo uma leve vantagem em relação ao terceiro cenário, a segunda configuração foi a utilizada como final.
